src/khepri_machine.erl

```diff
@@ -2322,3 +2340,126 @@ convert_state(State, 0, 1) ->
     State1 = list_to_tuple(Fields1),
     ?assert(is_state(State1)),
     State1.
+
+-spec update_projections(OldState, NewState) -> ok when
+      OldState :: khepri_machine:state(),
+      NewState :: khepri_machine:state().
+%% @doc Updates the machine's projections to account for changes between two
+%% states.
+%%
+%% This is used when installing a projection - the state will jump from the
+%% given `OldState' before the snapshot was installed to the given `NewState'
+%% after. When we swap states we need to update the projections: the records
+%% in the projection tables themselves but also which projection tables exist.
+%% The changes glossed over by the snapshot may include projection
+%% registrations and unregistrations so we need to initialize new projections
+%% and delete unregistered ones, and we need to ensure that the projection
+%% tables are up to date for any projections which didn't change.
+%%
+%% @private
+
+update_projections(OldState, NewState) ->
+    OldTree = get_tree(OldState),
+    OldProjections = set_of_projections(get_projections(OldState)),
+    NewTree = get_tree(NewState),
+    NewProjections = set_of_projections(get_projections(NewState)),
+
+    CommonProjections = sets:intersection(OldProjections, NewProjections),
+    DeletedProjections = sets:subtract(OldProjections, CommonProjections),
+    CreatedProjections = sets:subtract(NewProjections, CommonProjections),
+
+    %% Tear down any projections which were unregistered.
+    lists:foreach(
+      fun({_Pattern, Projection}) ->
+              _ = khepri_projection:delete(Projection)
+      end, sets:to_list(DeletedProjections)),
```

Comment:
Could the use of `sets:fold/3` be more efficient?

Reply:
I expect that we shouldn't have so many projections that it becomes noticeable but it does seem needless to me to convert to a list just for foreach. I switched to `sets:fold/3`.

Reply:
You are right, it will probably make little difference.